### ISSUE TITLE
Cluster Allocation Explain returns 400 when there are no unassigned shards to explain.

### ISSUE THREAD

**Issue by Mpdreamz** [elastic/elasticsearch#34800]
GET /_cluster/allocation/explain

returns a 400 with the following error 

```json
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "unable to find any unassigned shards to explain [ClusterAllocationExplainRequest[useAnyUnassignedShard=true,includeYesDecisions?=false]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "unable to find any unassigned shards to explain [ClusterAllocationExplainRequest[useAnyUnassignedShard=true,includeYesDecisions?=false]"
  },
  "status": 400
}
```

Should this return a `200` or possibly a `404` instead? cc @elastic/es-clients

**Comment by tabletick** [comment]
At least my contribution for the course in stockholm.. ;)

**Comment by jasontedor** [comment]
In general I think that indicating a client error is the correct response here, and to these types of situations. The server is indicating that it was asked to do something that does not make sense given the current state. I do not think that 404 is the correct response, as that could easily be mis-construed as there being a problem with the path sent to the server in the request.

**Comment by jpcarey** [comment]
@jasontedor it feels a bit weird to return a 400 on a GET request where no parameters were actually specified. It is supposed to return an explain for the allocation of the first unassigned shard. If no unassigned shards exist, why not respond with an empty body + 200? I fail to see why this GET should be categorized as a malformed request. A search request that results in `0` hits is not an error.

400 = `The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.`

**Comment by jasontedor** [comment]
I want to point out one thing here. This is a response to the comment about the intent of 400. I don't know there the quoted text is from. The specification for 400 is:

```
6.5.1.  400 Bad Request

   The 400 (Bad Request) status code indicates that the server cannot or
   will not process the request due to something that is perceived to be
   a client error (e.g., malformed request syntax, invalid request
   message framing, or deceptive request routing).
```

The key here is "perceived to be a client error". Malformed request syntax is only an example of what could be considered a bad request. The generic idea applies here though: asking for an explanation for an unassigned shard when there are no unassigned shards, we are considering this a client error. It's a bad request.

**Comment by jpcarey** [comment]
Grabbed from an older standard (rfc2616 10.4.1), sorry about that @jasontedor. Thank you for the clarification, although I'll still disagree (but accept the resolution here) 😄 My opinion, the client shouldn't need pre-existing knowledge of the servers "state". The response has confused me more than once, especially when the shard allocation state changed between the time checking cluster health and asking for explain.

**Comment by jasontedor** [comment]
> My opinion, the client shouldn't need pre-existing knowledge of the servers "state".

Given that the intended use of this API is explaining:
 - why a cluster is not green
 - why a shard is allocated where it is allocated

we think a client should.

> The response has confused me more than once, especially when the shard allocation state changed between the time checking cluster health and asking for explain.

I think this is an edge case.

**Comment by Mpdreamz** [comment]
Hate to be a stickler but I think the reasoning for `400` is quite a liberal interpretation of the RFC.

A `400` is widely accepted to mean: something you've sent does not validate (headers/payload) please go fix your request. 

Not send the request at a later time because we are not in a state to fulfill this request at this point of time.

The question this this API answers is give me the list of shards that are causing my cluster not be green. If there are no shards that cause it not to be green I think we should return a `200` with an empty array. 

This is what we do for e.g `GET /_cluster/pending_tasks` it will simply return an empty array if there are no tasks pending.

**Comment by karmi** [comment]
I agree with @Mpdreamz, and consider a `200` response appropriate in this situation.

In other words, when a client calls this API and gets an `200` status, and and empty JSON object/list, that's aligned with the purpose of the API call, in my understanding: no unassigned/inflight/etc shards, everything is OK.

**Comment by dakrone** [comment]
> A 400 is widely accepted to mean: something you've sent does not validate (headers/payload) please go fix your request.

That is correct, in this case, fix your request by specifying a shard to explain, regardless of the color of the cluster.

**Comment by HonzaKral** [comment]
Just to pile on - since there is *as of now* no way for the client code to reliably identify what is the underlying cause of a `400` (misshapen body VS illegal request)  then `400` is not usable from anyone's code as it requires a human to interpret.

**Comment by jasontedor** [comment]
This is an API for humans.

**Comment by Mpdreamz** [comment]
> This is an API for humans.

I don't think we should use this argument ever, we never communicated this explicitly nor should it change the design of how we expose API's. 

> That is correct, in this case, the request must be fixed by specifying a shard to explain, regardless of the color of the cluster.

The current behavior is to return only the first unallocated shard rather then listing all the unallocated shards.

Even a `404` would be better in this case with a body saying `found: false` for the same reason `GET /my-index/_doc/id-that-does-not-exist` does.
